source/nodejs/adaptivecards/src/markdown-formatter.ts

```diff
@@ -0,0 +1,182 @@
+/**
+ *  Markdown Formatter that supports bold, italic, hyperlinks, number lists and bullet lists.
+ *  It also gives user an option to add their own regex to format.
+ */
+
+class MarkdownFormatter {
+    text:string;
+    patterns:any[] = [];
+    styles:any[] = [];
+    patternTypes:any[] = [];
+    styleTypes:any[] = [];
+
+    static regexArray:any[];
+    static SPECIAL_CHAR_REGEX = new RegExp("[^|a-z\\\\s\\d]", 'gi');
+    
+	static MD_FORMATTER_CONFIG=[
```

Comment:
I have changed it according to this. But had to make the configurations as an array rather than JSON as we would have to overwrite the rule if the user passes the same type, say different pattern to format italic. Also, if it was JSON I would have to call the initialize logic mentioning every JSON key. It made sense to use an array. Please review the new code and let me know if it is fine.

Reply:
Thanks for making the changes.

I still believe that `configuration` should be declared as I suggested. Looking at your update, I can see that:
- There is still a `type` property, which is of type string
- That `type` property is used solely to allow the developer to replace the built-in rules, if the custom rules they pass have a type name that matches
- The only way for the developer to know what type names to use is either through documentation or by looking up the code

At the very least, `type` should be strongly typed, e.g.:
```typescript
export type BuiltInRuleTypeName = "orderedList" | "unorderedList" | ...;

export interface IFormatterRule {
    type?: BuiltInRuleTupeName;
    ...
}
```

But I do believe that's still less than optimal. With the approach I suggest, a developer can replace the rule for say `unorderedList` by doing this:
```typescript
MarkdownFormatter.configuration.unorderedList = { ... };
```

That's a super easy, strongly typed and discoverable way to replace a built-in rule IMO.

And adding custom rules is as easy as setting the `customRules` field:
```typescript
MarkdownFormatter.configuration.customRules = [
    {
        patternType: "start-end",
        ...
    }
]
```

Alternatively, `IFormatterConfiguration` could be declared like this:
```typescript
export interface IFormatterConfiguration {
    orderedList: IFromatterRule;
    unorderedList: IFormatterRule;
    bold: IFormatterRule;
    italic: IFormatterRule;
    hyperlink: IFormatterRule;
    [ruleName: string]: IFormatterStyleRule
}
```

which would allow developers to add/remove custom rules as follows, with the guarantee they can't conflict with a built-in rule:
```typescript
// Add a custom rule
MarkdownFormatter.configuration.myCustomRule = { ... };
// Remove a custom rule
delete MarkdownFormatter.configuration.myCustomRule;

```